src/sciline/pipeline.py

```diff
@@ -187,31 +173,40 @@ def _get_provider(
                 raise AmbiguousProvider("Multiple providers found for type", tp)
         raise UnsatisfiedRequirement("No provider found for type", tp)
 
-    def get(self, tp: Type[T], /) -> Delayed:
-        # We are slightly abusing Python's type system here, by using the
-        # self.get to get T, but actually it returns a Delayed that can
-        # compute T. We'd like to use Delayed[T], but that is not supported yet:
-        # https://github.com/dask/dask/pull/9256
-        #
-        # When building a workflow, there are two common problems:
-        #
-        # 1. Intermediate results are used more than once.
-        # 2. Intermediate results are large, so we generally do not want to keep them
-        #    in memory longer than necessary.
-        #
-        # To address these problems, we can internally build a graph of tasks, instead
-        # of directly creating dependencies between functions. Currently we use Dask
-        # for this. We cache call results to ensure Dask will recognize the task
-        # as the same object) and also wrap the function in dask.delayed.
-        if (cached := self._cache.get(tp)) is not None:
-            return cached
+    def build(self, tp: Type[T], /) -> Graph:
+        """
+        Return a dict of providers required for building the requested type `tp`.
+
+        The values are tuples containing the provider, the dict of bound typevars,
+        and the dict of arguments for the provider. The values in the latter dict
+        reference other keys in the returned graph.
 
+        Parameters
+        ----------
+        tp:
+            Type to build the graph for.
+        """
+        provider: Callable[..., T]
         provider, bound = self._get_provider(tp)
-        args = self._get_args(provider, bound=bound)
-        delayed = dask.delayed(provider)  # type: ignore[attr-defined]
-        return self._cache.setdefault(tp, delayed(**args))
+        tps = get_type_hints(provider)
+        args = {
+            name: _bind_free_typevars(t, bound=bound)
+            for name, t in tps.items()
+            if name != 'return'
+        }
+        graph: Graph = {tp: (provider, bound, args)}
+        for arg in args.values():
+            graph.update(self.build(arg))
```

Comment:
Potential infinite loop here (or up to max stack depth). @YooSunYoung has a solution based on `graphlib`.

Reply:
Do you mind just getting `RecursionError`?

Reply:
Yes. It would be nicer to report cycles in the graph more clearly. Maybe even during construction of the pipeline.